--- nmap-2.54BETA34/nbase/configure.in.orig	Sun Dec 30 12:54:38 2001
+++ nmap-2.54BETA34/nbase/configure.in	Wed May 22 00:33:23 2002
@@ -131,21 +131,22 @@
 AC_CHECK_FUNCS(vsnprintf snprintf asprintf asnprintf vasprintf vasnprintf,,
         [needsnprintf=yes])
 if test $needsnprintf = yes; then
-        LIBOBJS="$LIBOBJS snprintf.o"
+        AC_LIBOBJ([snprintf])
 fi
 
 AC_CHECK_FUNCS(usleep gettimeofday sleep, ,
-  [ LIBOBJS="$LIBOBJS nbase_time.o" ])
+  [ AC_LIBOBJ([nbase_time]) ])
 
 AC_REPLACE_FUNCS(inet_ntop inet_pton inet_aton)
 
 AC_CHECK_FUNC(getopt_long_only, , 
-  [ LIBOBJS="$LIBOBJS getopt.o getopt1.o" ])
+  [ AC_LIBOBJ([getopt]) ; AC_LIBOBJ([getopt1]) ])
 
 AC_CHECK_FUNCS(strcasecmp strncasecmp, , 
-  [ LIBOBJS="$LIBOBJS strcasecmp.o" ])
+  [ AC_LIBOBJ([strcasecmp]) ])
 
 dnl We always want some of our files
-LIBOBJS="$LIBOBJS nbase_str.o nbase_misc.o"
+AC_LIBOBJ([nbase_str])
+AC_LIBOBJ([nbase_misc])
 
 AC_OUTPUT(Makefile)
