--- nmap-3.20/nmapfe.desktop~	Thu Nov 14 11:37:13 2002
+++ nmap-3.20/nmapfe.desktop	Fri Mar 28 18:09:13 2003
@@ -1,6 +1,14 @@
 [Desktop Entry]
-Name=Nmapfe
+Encoding=UTF-8
+Name=Nmap
 Comment=Gtk+ frontend to nmap
+Comment[es]=Herramienta de exploración de la rede y seguridad
+Comment[pl]=Frontend gtk+ do nmap
+Comment[pt]=Ferramenta de exploração da rede e segurança
+Comment[ru]=Утилита сканирования сети и аудита безопасности
+Comment[uk]=Утиліта сканування мережі та аудиту безпеки
 Exec=nmapfe
+Icon=nmap.png
 Terminal=0
 Type=Application
+Categories=Application;Network
