diff -burN nmap-3.46.orig/Makefile.in nmap-3.46/Makefile.in
--- nmap-3.46.orig/Makefile.in	2003-09-27 13:17:01.000000000 +0200
+++ nmap-3.46/Makefile.in	2003-09-27 13:20:19.000000000 +0200
@@ -254,19 +254,19 @@
 	$(MAKEDEPEND) $(INCLS) -s "# DO NOT DELETE" -- $(DEFS) -- $(SRCS)
 
 install-nmap: $(TARGET)
-	$(SHTOOL) mkdir -f -p -m 755 $(bindir) $(mandir)/man1 $(nmapdatadir)
-	$(INSTALL) -c -m 755 nmap $(bindir)/nmap
-	$(INSTALL) -c -m 644 docs/$(TARGET).1 $(mandir)/man1/$(TARGET).1
-	$(INSTALL) -c -m 644 nmap-os-fingerprints  $(nmapdatadir)/nmap-os-fingerprints
-	$(INSTALL) -c -m 644 nmap-service-probes  $(nmapdatadir)/nmap-service-probes
-	$(INSTALL) -c -m 644 nmap-services  $(nmapdatadir)/nmap-services
-	$(INSTALL) -c -m 644 nmap-protocols  $(nmapdatadir)/nmap-protocols
-	$(INSTALL) -c -m 644 nmap-rpc  $(nmapdatadir)/nmap-rpc
+	$(SHTOOL) mkdir -f -p -m 755 $(DESTDIR)$(bindir) $(DESTDIR)$(mandir)/man1 $(DESTDIR)$(nmapdatadir)
+	$(INSTALL) -c -m 755 nmap $(DESTDIR)$(bindir)/nmap
+	$(INSTALL) -c -m 644 docs/$(TARGET).1 $(DESTDIR)$(mandir)/man1/$(TARGET).1
+	$(INSTALL) -c -m 644 nmap-os-fingerprints  $(DESTDIR)$(nmapdatadir)/nmap-os-fingerprints
+	$(INSTALL) -c -m 644 nmap-service-probes  $(DESTDIR)$(nmapdatadir)/nmap-service-probes
+	$(INSTALL) -c -m 644 nmap-services  $(DESTDIR)$(nmapdatadir)/nmap-services
+	$(INSTALL) -c -m 644 nmap-protocols  $(DESTDIR)$(nmapdatadir)/nmap-protocols
+	$(INSTALL) -c -m 644 nmap-rpc  $(DESTDIR)$(nmapdatadir)/nmap-rpc
 
 install-nmapfe: $(TARGETNMAPFE)
-	$(SHTOOL) mkdir -f -p -m 755 $(bindir) $(mandir)/man1 $(nmapdatadir) $(deskdir)
+	$(SHTOOL) mkdir -f -p -m 755 $(DESTDIR)$(bindir) $(DESTDIR)$(mandir)/man1 $(DESTDIR)$(nmapdatadir) $(DESTDIR)$(deskdir)
 	@echo "If the next command fails -- you cannot use the X front end"
-	-test -f nmapfe/nmapfe && $(INSTALL) -c -m 755 nmapfe/nmapfe $(bindir)/nmapfe && rm -f $(bindir)/xnmap && $(SHTOOL) mkln -f -s $(bindir)/nmapfe $(bindir)/xnmap && $(INSTALL) -c -m 644 nmapfe.desktop $(deskdir)/nmapfe.desktop && $(INSTALL) -c -m 644 docs/nmapfe.1 $(mandir)/man1/nmapfe.1 && $(INSTALL) -c -m 644 docs/xnmap.1 $(mandir)/man1/xnmap.1
+	-test -f nmapfe/nmapfe && $(INSTALL) -c -m 755 nmapfe/nmapfe $(DESTDIR)$(bindir)/nmapfe && rm -f $(DESTDIR)$(bindir)/xnmap && $(SHTOOL) mkln -f -s $(DESTDIR)$(bindir)/nmapfe $(DESTDIR)$(bindir)/xnmap && $(INSTALL) -c -m 644 nmapfe.desktop $(DESTDIR)$(deskdir)/nmapfe.desktop && $(INSTALL) -c -m 644 docs/nmapfe.1 $(DESTDIR)$(mandir)/man1/nmapfe.1 && $(INSTALL) -c -m 644 docs/xnmap.1 $(DESTDIR)$(mandir)/man1/xnmap.1
 
 install: install-nmap $(INSTALLNMAPFE)
 
